File: 2019/elixir/lib/Aoc2019/Day02.ex
defmodule Aoc2019.Day02 do
  def part_one(input, noun \\ 12, verb \\ 2) do
    input
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
    |> List.replace_at(1, noun)
    |> List.replace_at(2, verb)
    |> process
    |> Enum.at(0)
  end

  def process(list, pointer \\ 0) do
    process(list, pointer, Enum.slice(list, pointer, 4))
  end

  defp process(list, _, [99 | _]), do: list

  defp process(list, pointer, [opcode, a, b, r]) do
    list
    |> List.replace_at(r, Enum.reduce([Enum.at(list, a), Enum.at(list, b)], op(opcode)))
    |> process(pointer + 4)
  end

  defp op(1), do: &+/2
  defp op(2), do: &*/2

  def part_two(input) do
    try do
      for noun <- 0..99,
          verb <- 0..99,
          do: if(part_one(input, noun, verb) == 19_690_720, do: throw(noun * 100 + verb))
    catch
      x -> x
    end
  end
end

File: 2019/ruby/lib/aoc2019/day02.rb
module Aoc2019
  class Day02
    class Program
      attr_accessor :state, :p, :finished

      def initialize(*state)
        @state = state
        @finished = false
        @p = 0
      end

      def step
        case state[@p]
        when 1 then state[state[@p + 3]] = state[state[@p + 1]] + state[state[@p + 2]]
        when 2 then state[state[@p + 3]] = state[state[@p + 1]] * state[state[@p + 2]]
        when 99 then @finished = true
        else raise "Unknown opcode! #{@p}: #{@state}"
        end
        @p += 4
      end

      def run
        self.step until @finished
        self
      end
    end

    def self.part_one(input)
      program = Program.new(*input.split(",").map(&:to_i))
      program.state[1] = 12
      program.state[2] = 2
      program.run.state[0]
    end

    def self.part_two(input)
      catch(:complete) do
        0.upto(99) do |noun|
          0.upto(99) do |verb|
            program = Program.new(*input.split(",").map(&:to_i))
            program.state[1] = noun
            program.state[2] = verb
            program.run.state[0]
            throw(:complete, 100 * noun + verb) if program.state[0] == 19690720
          end
        end
      end
    end
  end
end
